diff -r 8dbea571bd08 d/d-lang.cc
--- a/d/d-lang.cc	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/d-lang.cc	Thu Dec 02 12:50:13 2010 -0500
@@ -375,6 +375,9 @@
     maybe_fixup_cygwin();
 
     VersionCondition::addPredefinedGlobalIdent("GNU");
+#ifdef __MINGW32__
+    VersionCondition::addPredefinedGlobalIdent("MinGW");
+#endif
 #if V2
     VersionCondition::addPredefinedGlobalIdent("D_Version2");
 #endif
diff -r 8dbea571bd08 d/dmd2/root.c
--- a/d/dmd2/root.c	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/dmd2/root.c	Thu Dec 02 12:50:13 2010 -0500
@@ -1891,8 +1891,6 @@
             psize = count + 1;
         else
             break;
-#else
-    assert(0);
 #endif
         p = (dchar *) alloca(psize * 2);        // buffer too small, try again with larger size
     }
diff -r 8dbea571bd08 d/druntime/object_.d
--- a/d/druntime/object_.d	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/druntime/object_.d	Thu Dec 02 12:50:13 2010 -0500
@@ -1699,6 +1699,18 @@
 // Posix: this gets initialized in _moduleCtor()
 extern (C) __gshared ModuleInfo*[] _moduleinfo_array;
 
+version (MinGW)
+{
+    // This linked list is created by a compiler generated function inserted
+    // into the .ctor list by the compiler.
+    struct ModuleReference
+    {
+        ModuleReference* next;
+        ModuleInfo*      mod;
+    }
+
+    extern (C) __gshared ModuleReference* _Dmodule_ref;   // start of linked list
+}
 
 version (linux)
 {
@@ -1764,6 +1776,22 @@
 extern (C) void _moduleCtor()
 {
     debug(PRINTF) printf("_moduleCtor()\n");
+	
+    version (MinGW)
+    {
+        int len = 0;
+        ModuleReference *mr;
+
+        for (mr = _Dmodule_ref; mr; mr = mr.next)
+            len++;
+        _moduleinfo_array = new ModuleInfo*[len];
+        len = 0;
+        for (mr = _Dmodule_ref; mr; mr = mr.next)
+        {   _moduleinfo_array[len] = mr.mod;
+            len++;
+        }
+    }
+	
     version (linux)
     {
         int len = 0;
diff -r 8dbea571bd08 d/druntime/rt/dmain2.d
--- a/d/druntime/rt/dmain2.d	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/druntime/rt/dmain2.d	Thu Dec 02 12:50:13 2010 -0500
@@ -271,6 +271,11 @@
         _STI_monitor_staticctor();
         _STI_critical_init();
     }
+	version (MinGW)
+	{
+        _STI_monitor_staticctor();
+        _STI_critical_init();
+	}
 }
 
 alias void delegate(Throwable) ExceptionHandler;
@@ -283,7 +288,7 @@
     {
         gc_init();
         initStaticDataGC();
-        version (Windows)
+        version (Windows) version (MinGW) {} else
             _minit();
         _moduleCtor();
         _moduleTlsCtor();
@@ -310,6 +315,11 @@
         _STD_critical_term();
         _STD_monitor_staticdtor();
     }
+    version (MinGW)
+    {
+        _STD_critical_term();
+        _STD_monitor_staticdtor();
+    }
 }
 
 extern (C) bool rt_term(ExceptionHandler dg = null)
@@ -386,6 +396,12 @@
         _STI_critical_init();
     }
 
+    version (MinGW)
+    {
+        _STI_monitor_staticctor();
+        _STI_critical_init();
+    }
+	
     version (Windows)
     {
         wchar_t*  wcbuf = GetCommandLineW();
@@ -497,7 +513,7 @@
     {
         gc_init();
         initStaticDataGC();
-        version (Windows)
+        version (Windows) version (MinGW) {} else
             _minit();
         _moduleCtor();
         _moduleTlsCtor();
@@ -519,5 +535,12 @@
         _STD_critical_term();
         _STD_monitor_staticdtor();
     }
+
+    version (MinGW)
+    {
+        _STD_critical_term();
+        _STD_monitor_staticdtor();
+    }
+
     return result;
 }
diff -r 8dbea571bd08 d/druntime/rt/memory.d
--- a/d/druntime/rt/memory.d	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/druntime/rt/memory.d	Thu Dec 02 12:50:13 2010 -0500
@@ -155,7 +155,26 @@
 
 private
 {
-    version( Windows )
+    version( MinGW )
+    {
+        extern (C)
+        {
+            extern __gshared
+            {
+                int _data_start__;
+                int _data_end__;
+                int _bss_start__;
+                int _bss_end__;
+
+                alias _data_start__ Data_Start;
+                alias _data_end__ Data_End;
+				
+                alias _bss_start__ Data_Start2;
+                alias _bss_end__ Data_End2;
+			}
+		}
+	}
+    else version( Windows )
     {
         extern (C)
         {
@@ -218,7 +237,11 @@
 
 void initStaticDataGC()
 {
-    version( Windows )
+    version( MinGW )
+    {
+        gc_addRange( &Data_Start, cast(size_t) &Data_End2 - cast(size_t) &Data_Start );	
+    }
+    else version( Windows )
     {
         gc_addRange( &_xi_a, cast(size_t) &_end - cast(size_t) &_xi_a );
     }
diff -r 8dbea571bd08 d/phobos2/Makefile.am
--- a/d/phobos2/Makefile.am	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/phobos2/Makefile.am	Thu Dec 02 12:50:13 2010 -0500
@@ -259,7 +259,7 @@
 WINDOWS_OBJS=std/c/windows/com.o std/c/windows/stat.o std/c/windows/windows.o \
              std/c/windows/winsock.o std/windows/charset.o \
              std/windows/iunknown.o std/windows/registry.o \
-             std/windows/syserror.o
+             std/windows/syserror.o std/__fileinit.o
 
 D_EXTRA_OBJS=@D_EXTRA_OBJS@
 
diff -r 8dbea571bd08 d/phobos2/Makefile.in
--- a/d/phobos2/Makefile.in	Wed Dec 01 22:57:23 2010 +0000
+++ b/d/phobos2/Makefile.in	Thu Dec 02 12:50:13 2010 -0500
@@ -363,7 +363,7 @@
 WINDOWS_OBJS = std/c/windows/com.o std/c/windows/stat.o std/c/windows/windows.o \
              std/c/windows/winsock.o std/windows/charset.o \
              std/windows/iunknown.o std/windows/registry.o \
-             std/windows/syserror.o
+             std/windows/syserror.o std/__fileinit.o
 
 ALL_PHOBOS_OBJS = $(D_EXTRA_OBJS) $(MAIN_OBJS) $(ZLIB_OBJS)
 
