diff -r 106a741599c6 d/phobos2/gcc/deh.d
--- a/d/phobos2/gcc/deh.d	Tue Nov 23 15:02:08 2010 +0000
+++ b/d/phobos2/gcc/deh.d	Wed Nov 24 17:58:50 2010 +0000
@@ -28,6 +28,12 @@
 private import core.memory;
 private import core.stdc.stdlib;
 
+extern (C)
+{
+    void _d_setUnhandled(Object *);
+    void _d_createTrace(Object *);
+}
+
 static if (Use_ARM_EABI_Unwinder)
     const _Unwind_Exception_Class GDC_Exception_Class =
         ['G','N','U','C',
@@ -153,6 +159,10 @@
         exc.unwindHeader.exception_class[] = GDC_Exception_Class[];
     exc.unwindHeader.exception_cleanup = & _gdc_cleanupException;
 
+    // Runtime now expects us to do this first before unwinding.
+    _d_createTrace (&exc.obj);
+    _d_setUnhandled (&exc.obj);
+
     version (GNU_SjLj_Exceptions) {
         _Unwind_SjLj_RaiseException (&exc.unwindHeader);
     } else {
